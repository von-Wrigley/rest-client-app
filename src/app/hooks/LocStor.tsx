'use client';

import { useState, useEffect } from 'react';

function getStorageValue(key: string, defaultValue: []) {
  if (typeof window !== 'undefined') {
    const saved = localStorage.getItem(key);
    const initial = JSON.parse(saved as string);
    return initial || defaultValue;
  }
}

export const useLocalStorage = (key: string, defaultValue: []) => {
  const [value, setValue] = useState(() => {
    if (typeof window !== 'undefined') {
      return getStorageValue(key, defaultValue);
    }
  });

  useEffect(() => {
    localStorage.setItem(key, JSON.stringify(value));
  }, [key, value]);

  return [value, setValue];
};
