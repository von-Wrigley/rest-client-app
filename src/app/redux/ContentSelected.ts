import { createSlice, type PayloadAction } from "@reduxjs/toolkit";

type ResponseDet = {
  resOk: boolean;
  resStatus: number;
};

type InitState = {
  method: string;
  inputURL: string;
  bodyReq: string;
  bodyRes: string;
  details: ResponseDet[];
  generateCode: string[];
  selectedLang: string;
  headers: Record<string, string>[];
};

type State = {
  selectedContent: InitState;
};

const initialState: State = {
  selectedContent: {
    method: "",
    inputURL: "",
    bodyReq: "",
    bodyRes: "",
    details: [],
    generateCode: [],
    selectedLang: "",
    headers: [],
  },
};

const StoreSelectedSlice = createSlice({
  name: "selectedItems",
  initialState,
  reducers: {
    addMethod: (state, action: PayloadAction<string>) => {
      state.selectedContent.method = action.payload;
    },
    addInputURL: (state, action: PayloadAction<string>) => {
      state.selectedContent.inputURL = action.payload;
    },
    addBodyReq: (state, action: PayloadAction<string>) => {
      state.selectedContent.bodyReq = action.payload;
    },
    addBodyRes: (state, action: PayloadAction<string>) => {
      state.selectedContent.bodyRes = action.payload;
    },
    resContentDetails: (state, action: PayloadAction<ResponseDet[]>) => {
      state.selectedContent.details = action.payload;
    },
    generateCodePOSTMAN: (state, action: PayloadAction<string[]>) => {
      state.selectedContent.generateCode = action.payload;
    },
    selectedLanguage: (state, action: PayloadAction<string>) => {
      state.selectedContent.selectedLang = action.payload;
    },
    selectedHeaders: (
      state,
      action: PayloadAction<Record<string, string>[]>
    ) => {
      state.selectedContent.headers = action.payload;
    },
  },
});
export const {
  addMethod,
  addInputURL,
  addBodyReq,
  addBodyRes,
  resContentDetails,
  generateCodePOSTMAN,
  selectedLanguage,
  selectedHeaders,
} = StoreSelectedSlice.actions;

export default StoreSelectedSlice.reducer;
