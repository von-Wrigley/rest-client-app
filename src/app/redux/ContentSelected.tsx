import { createSlice, type PayloadAction  } from "@reduxjs/toolkit"

type ResponseDet = {
  resOk: boolean;
  resStatus: number;
};


type InitSt = {
  method: string,
  inputURL: string, 
  bodyReq: string,
  bodyRes: string,
  details: ResponseDet[],
  generateCode: [], 
  selectedLang: string,
  headers: []
}


const initState = {
  selectedContent: {method: '', inputURL: '', bodyReq: '', bodyRes: '', details: [], generateCode: [], selectedLang: "", headers: []}
}

const StoreSelecedSlice = createSlice({
    name: 'selectedItems',
    initialState: initState,
    reducers: {
        addItem: (state, action: PayloadAction) => {
         state.selectedContent.method = action.payload
                 
        },
        addInputURL: (state, action: PayloadAction) => {
          state.selectedContent.inputURL = action.payload
                  
         },
         addBodyReq: (state, action: PayloadAction) => {
          state.selectedContent.bodyReq = action.payload
                  
         },
         addBodyRes: (state, action: PayloadAction) => {
          state.selectedContent.bodyRes = action.payload
                  
         },
         resContentDetails: (state, action: PayloadAction) => {
          state.selectedContent.details =action.payload
                  
         },
         generateCodePOSTMAN: (state, action: PayloadAction) => {
          state.selectedContent.generateCode = action.payload 
                  
         },
         selectedLanguage: (state, action: PayloadAction) => {
          state.selectedContent.selectedLang =action.payload
                  
         },
         selectedHeaders: (state, action: PayloadAction) => {
          state.selectedContent.headers =action.payload
                  
         },
       
       
   
    }
})
export const {addItem, addInputURL, addBodyReq, addBodyRes,resContentDetails, generateCodePOSTMAN, selectedLanguage, selectedHeaders } = StoreSelecedSlice.actions
export default StoreSelecedSlice.reducer


