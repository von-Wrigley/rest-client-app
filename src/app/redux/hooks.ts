import { useDispatch, useSelector } from "react-redux";
import type { AppDispatch, RootState } from "./store";

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector = useSelector.withTypes<RootState>();
