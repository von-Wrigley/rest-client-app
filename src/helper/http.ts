import { HttpMethod } from "@/app/types/http";
import { headers } from "next/headers";
import { NextResponse } from "next/server";

export const proxyResponse = async (backendResponse: Response) => {
  const contentType =
    backendResponse.headers.get("Content-Type") ?? "text/plain";
  const body = await backendResponse.arrayBuffer();

  return new NextResponse(body, {
    status: backendResponse.status,
    headers: {
      "Content-Type": contentType,
      "X-Status-Text": backendResponse.statusText,
    },
  });
};

export const proxyFetch = async (
  method: HttpMethod,
  params: Promise<{ slug: string[] }>,
  options?: globalThis.RequestInit,
) => {
  const { slug } = await params;
  const newSlug = slug.join("/");
  const headersList = await headers();
  const rawHeader = headersList.get("X-Custom-Headers");
  const parsed: { name: string; value: string }[] = rawHeader
    ? JSON.parse(rawHeader)
    : [];
  const extraHeaders = Object.fromEntries(
    parsed.map(({ name, value }) => [name, value]),
  );

  const backendResponse = await fetch(newSlug, {
    method,
    ...extraHeaders,
    ...options,
  });

  return backendResponse;
};
