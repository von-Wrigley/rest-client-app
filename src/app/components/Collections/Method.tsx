"use client";
import React, { useEffect, useState } from "react";
import { addItem } from "@/app/redux/ContentSelected";
import { useAppDispatch, useAppSelector } from "@/app/redux/hooks";
import { usePathname } from "next/navigation";
import { useParams } from "next/navigation";

function Method() {
  const params = useParams();
  const inputfromhistory = params.slug?.[0];
  const inputURLHistory = params.slug?.[1];

  const [stateMethod, setMethod] = useState<string>(inputfromhistory ?? "");
  const dispatch = useAppDispatch();
  const pathname = usePathname();
  const inputURL = useAppSelector(
    (state) => state.selected.selectedContent,
  ).inputURL;
  const bodyReq2 = useAppSelector(
    (state) => state.selected.selectedContent,
  ).bodyReq;
  const headers = useAppSelector(
    (state) => state.selected.selectedContent,
  ).headers;

  const handlechange = ({ target }: React.ChangeEvent<HTMLSelectElement>) => {
    setMethod(target.value);
  };

  useEffect(() => {
    dispatch(addItem(stateMethod));

    if (inputfromhistory) {
      window.history.replaceState(
        null,
        "",
        "/personal" +
          +(stateMethod.length > 0 ? "/" + stateMethod : "") +
          "/" +
          inputURLHistory +
          "?" +
          headers,
      );
    }

    window.history.replaceState(
      window.history.state,
      "",
      pathname +
        (stateMethod.length > 0 ? "/" + stateMethod : "") +
        (inputURL.length > 0 ? "/" + inputURL : "") +
        (bodyReq2.length > 0 ? "/" + bodyReq2 : "") +
        "?" +
        headers,
    );
  }, [stateMethod]);

  return (
    <select
      name="Method"
      id="method"
      className="p-2 m-2"
      value={stateMethod}
      onChange={handlechange}
    >
      <option disabled value="">
        Select Method
      </option>
      <option value="GET">GET</option>
      <option value="DELETE">DELETE</option>
      <option value="POST">POST</option>
      <option value="PUT">PUT</option>
      <option value="PATCH">PATCH</option>
    </select>
  );
}

export default Method;
