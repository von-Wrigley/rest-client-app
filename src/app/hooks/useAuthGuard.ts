import { useEffect, useState } from "react";
import { useRouter } from "next/navigation";
import { supabase } from "@/helper/supabaseClient";

export const useAuthGuard = () => {
  const router = useRouter();
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    const checkAuth = async () => {
      try {
        const {
          data: { session },
        } = await supabase.auth.getSession();

        if (!session) {
          router.push("/unauthorized");
        } else {
          setIsLoading(false);
        }
      } catch (error) {
        console.error(error);
        router.push("/unauthorized");
      }
    };

    checkAuth();
  }, [router]);

  return isLoading;
};
