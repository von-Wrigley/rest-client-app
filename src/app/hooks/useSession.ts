import { useEffect, useState } from "react";
import { Session } from "@supabase/supabase-js";
import { supabase } from "@/helper/supabaseClient";

export function useSession() {
  const [session, setSession] = useState<Session | null>(null);

  useEffect(() => {
    supabase.auth.getSession().then(({ data: { session } }) => {
      setSession(session);
    });

    const {
      data: { subscription },
    } = supabase.auth.onAuthStateChange((_event, session) => {
      setSession(session);
    });

    return () => subscription.unsubscribe();
  }, []);

  return session;
}
