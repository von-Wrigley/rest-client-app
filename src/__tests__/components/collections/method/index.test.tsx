import Method from "@/components/collections/method";
import { fireEvent, render, screen } from "@testing-library/react";
import { useTranslations } from "next-intl";
import { HttpMethod } from "@/app/types/http";
import { addMethod } from "@/app/redux/ContentSelected";
import { useAppDispatch } from "@/app/redux/hooks";

jest.mock("next-intl", () => ({
  useTranslations: jest.fn(),
}));

jest.mock("@/app/redux/hooks", () => ({
  useAppDispatch: jest.fn(),
  useAppSelector: jest.fn(),
}));

jest.mock("@/app/redux/ContentSelected", () => ({
  addMethod: jest.fn(),
}));

describe("Method Componeent", () => {
  const mockT = jest.fn();
  const mockDispatch = jest.fn();

  beforeEach(() => {
    (useTranslations as jest.Mock).mockReturnValue(mockT);
    (useAppDispatch as jest.Mock).mockReturnValue(mockDispatch);
    mockT.mockImplementation((key) => {
      const translations: Record<string, string> = {
        SelectMethod: "Select Method",
      };
      return translations[key];
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders default option in select", () => {
    render(<Method />);

    const selectComp = screen.getByRole("combobox");
    const defaultOption = screen.getByText("Select Method");

    expect(selectComp).toBeInTheDocument();
    expect(defaultOption).toBeDisabled();
    expect(defaultOption).toHaveValue("");
  });
  it("show current option", () => {
    render(<Method />);

    const selectComp = screen.getByRole("combobox");
    expect(selectComp).toHaveValue("GET");
  });

  it("renders methods as options", () => {
    render(<Method />);

    Object.values(HttpMethod).forEach((x) => {
      expect(screen.getByText(x)).toBeInTheDocument();
    });
  });

  it("dispatch value on Select change", () => {
    render(<Method />);

    const selectComp = screen.getByRole("combobox");
    fireEvent.change(selectComp, { target: { value: "PUT" } });

    expect(addMethod).toHaveBeenCalledWith("PUT");
    expect(mockDispatch).toHaveBeenCalledTimes(1);
  });

  it("match snapshot", () => {
    const { asFragment } = render(<Method />);
    expect(asFragment()).toMatchSnapshot();
  });
});
