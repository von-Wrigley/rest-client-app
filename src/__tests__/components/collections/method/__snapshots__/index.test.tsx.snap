// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Method Componeent match snapshot 1`] = `
<DocumentFragment>
  <select
    class="method"
    id="method"
    name="Method"
  >
    <option
      disabled=""
      value=""
    >
      Select Method
    </option>
    <option
      value="GET"
    >
      GET
    </option>
    <option
      value="POST"
    >
      POST
    </option>
    <option
      value="PUT"
    >
      PUT
    </option>
    <option
      value="DELETE"
    >
      DELETE
    </option>
    <option
      value="PATCH"
    >
      PATCH
    </option>
  </select>
</DocumentFragment>
`;
